### PR TITLE
Implement a passive check for cookies without SameSite set

### PR DESCRIPTION
**Pull request by psiinon**
Fixes zaproxy/zaproxy#2716 Implement a passive check for cookies without SameSite set

### CODE REVIEW

**Comment by psiinon** [comment]
Repushed with fix and unit tests

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+		
+	}
+	
+	private void raiseAlert(HttpMessage msg, int id, String name, String description) {
+		String evidence = name;
+		// Try to find the exact Set-Cookie string
+		Pattern pattern = Pattern.compile("Set-Cookie.*" + name, Pattern.CASE_INSENSITIVE);
+		Matcher matcher = pattern.matcher(msg.getResponseHeader().toString());
+		if (matcher.find()) {
+			evidence = matcher.group();
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java]: Couldn't there be multiple matches at this point? So perhaps you want `group(1)`?

Comment (psiinon) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java]: It should just return the next match, ie the same as group(1), shouldnt it?

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import net.htmlparser.jericho.Source;
+
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpHeader;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.pscan.PassiveScanThread;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java]: This extra line could be removed.

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		}
+
+		Alert alert = new Alert(getPluginId(), Alert.RISK_LOW, Alert.CONFIDENCE_MEDIUM, 
+		    	getName());
+		    	alert.setDetail(
+		    		description, 
+		    		msg.getRequestHeader().getURI().toString(),
+		    		name, "", "",
+		    		getSolution(), 
+		            getReference(),
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/CookieSameSiteScanner.java]: Tabbing or spacing seems to have changed here...

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties [hunk 4]
```diff
@@ -1,6 +1,14 @@
 # This file defines the default (English) variants of all of the internationalised messages
 pscanalpha.desc = Passive Scan Rules - alpha
 
+pscanalpha.cookiesamesitescanner.name = Cookie Without SameSite Attribute
+pscanalpha.cookiesamesitescanner.desc=A cookie has been set without the SameSite attribute, which means that the cookie can be sent as a result of a 'cross-site' request. \
+The SameSite attribute is an effective counter measure to cross-site request forgery, cross-site script inclusion and timing attacks.
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties]: I'm a fan of the oxford or serial comma. Not sure what you guys think, but I'd suggest one following "script inclusion"....just my 2cents.

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties [hunk 5]
```diff
@@ -1,6 +1,14 @@
...
+pscanalpha.cookiesamesitescanner.badval=A cookie has been set with an invalid SameSite attribute value, which means that the cookie can be sent as a result of a 'cross-site' request. \
+The SameSite attribute is an effective counter measure to cross-site request forgery, cross-site script inclusion and timing attacks.
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties]: Same as above, re: commas.

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties [hunk 6]
```diff
@@ -1,6 +1,14 @@
...
+pscanalpha.cookiesamesitescanner.soln=Ensure that the SameSite attribute is set to either 'lax' or ideally 'strict' for all cookies.
+pscanalpha.cookiesamesitescanner.refs=https://tools.ietf.org/html/draft-west-first-party-cookies-07
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties]: I think this recently changed name: https://tools.ietf.org/html/draft-ietf-httpbis-cookie-same-site-00

I'm not really use how the RFC process does naming/version control, but the one linked above is dated more recently...

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties (hunk 4 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties]: I'm fine with that (just need to learn when/how to use it ;)

File: src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties (hunk 6 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties]: Maybe just:
https://tools.ietf.org/html/draft-ietf-httpbis-cookie-same-site

which is redirected to that one (latest revision?).

**Comment by psiinon** [comment]
Hopefully fixed all of those comments

**Comment by thc202** [comment]
Looks good to me.

**Comment by kingthorin** [comment]
Thanks!

**Pull request merged**